composer/datasets/streaming/format.py

```diff
@@ -0,0 +1,237 @@
+    ends = num_values * np.int64().nbytes + sizes.cumsum()
+    begins = ends - sizes
+    values = []
+    for begin, end in zip(begins, ends):
+        value = data[begin:end]
+        values.append(value)
+    return dict(zip(keys, values))
+
+
+def read_array(fp: BufferedIOBase, count: int, dtype: DTypeLike) -> NDArray:
+    """Load the count items from the file handle, advancing its position.
+
+    Args:
+        fp (BufferedIOBase): File handle.
+        count (int): Number of items to read.
+        dtype (np.dtype): Dtype of the items.
+
+    Returns:
+        np.ndarray: The read array.
+    """
+    num_bytes = count * dtype().nbytes  # type: ignore
+    data = fp.read(num_bytes)
+    return np.frombuffer(data, dtype)
+
+
+class StreamingDatasetIndex(object):
+    """Streaming dataset index file, containing all the info about shards.
+
+    The shards are binary buffers with samples concatenated together. All the
+    offset info across the whole dataset is contained in the index file. Workers
+    read this file to calculate how much of which shards their slice is.
+
+    Each sample is a dict of str to bytes. All samples must contain the same
+    dict keys (fields). These strings are stored in the index file for
+    efficiency.
+    """
+
+    def __init__(self, samples_per_shard: Union[NDArray, Sequence[int]], bytes_per_shard: Union[NDArray, Sequence[int]],
+                 bytes_per_sample: Union[NDArray, Sequence[int]], fields: Sequence[str]) -> None:
+        """Initialize with sample statistics.
+
+        Args:
+            samples_per_shard (NDArray): Number of samples of each shard.
+            bytes_per_shard (NDArray): Size in bytes of each shard.
+            bytes_per_sample (NDArray): Size in bytes of each sample across all shards.
+            fields (Sequence[str]): The names of the samples' fields in order.
+        """
+        self.samples_per_shard = np.asarray(samples_per_shard, np.int64)
+        self.bytes_per_shard = np.asarray(bytes_per_shard, np.int64)
+        self.bytes_per_sample = np.asarray(bytes_per_sample, np.int64)
+        self.fields = fields
+
+        # Totals.
+        self.num_shards = len(samples_per_shard)
+        self.total_bytes = sum(bytes_per_shard)
+        self.total_samples = len(bytes_per_sample)
+        self.num_fields = len(fields)
+
+        # Shard -> sample range.
+        self.shard_ends = self.samples_per_shard.cumsum()
+        self.shard_begins = self.shard_ends - self.samples_per_shard  # type: ignore
+
+        # Sample -> shard, byte offset within shard.
+        self.sample_shards, self.sample_shard_offsets = self.locate_samples()
+
+    @classmethod
+    def loads(cls, data: bytes):
+        """Load a StreamingDatasetIndex from raw bytes.
+
+        Args:
+            data (bytes): The serialized form.
+
+        Returns:
+            cls: The loaded object.
+        """
+        fp = BytesIO(data)
```

Comment:
Tracking this but not addressing in this PR.

Reply:
I think this would be pretty simple to fix? Just move the logic here to `load` below, and have `loads` call `load` (rather than the other way around).

Reply:
Done.